test/gtest/ucp/test_ucp_proto_mock.cc

```diff
@@ -0,0 +1,408 @@
+                                 << proto_attr.config;
+            }
+
+            range_end = proto_attr.max_msg_length;
+        } while (range_end != SIZE_MAX);
+    }
+
+    static void
+    check_proto_select_elem(ucp_worker_h worker,
+                            const ucp_proto_select_elem_t &select_elem,
+                            const proto_select_data_vec_t &data)
+    {
+        for (auto &it : data) {
+            size_t start = string_to_memunits(it.range_start);
+            EXPECT_TRUE(select_elem_match(worker, select_elem, it, start));
+            /* As we cannot get range_start directly, we assert that protocol
+             * is different at that range */
+            if (start > 0) {
+                EXPECT_FALSE(
+                        select_elem_match(worker, select_elem, it, start - 1));
+            }
+        }
+    }
+
+    static void check_proto_select(ucp_worker_h worker,
+                                   const ucp_proto_select_t &proto_select,
+                                   const proto_select_data_vec_t &data,
+                                   ucp_proto_select_key_t key)
+    {
+        ucp_proto_select_elem_t select_elem;
+        ucp_proto_select_key_t select_key;
+
+        kh_foreach(proto_select.hash, select_key.u64, select_elem,
+            if (key_match(key, select_key)) {
+                check_proto_select_elem(worker, select_elem, data);
+            });
+
+        if (testing::Test::HasFailure()) {
+            kh_foreach(proto_select.hash, select_key.u64, select_elem,
+                if (key_match(key, select_key)) {
+                    UCS_TEST_MESSAGE << "Key op flags: "
+                                     << (int)select_key.param.op_id_flags
+                                     << ", attr: "
+                                     << (int)select_key.param.op_attr;
+                    dump_select_elem(worker, select_elem);
+                });
+        }
+    }
+
+    static bool
+    key_match(ucp_proto_select_key_t req, ucp_proto_select_key_t actual)
+    {
+#define CMP_FIELD(FIELD) \
+        if ((UINT8_MAX != req.param.FIELD) && \
+            (req.param.FIELD != actual.param.FIELD)) { \
+            return false; \
+        }
+
+        CMP_FIELD(op_id_flags);
+        CMP_FIELD(op_attr);
+        CMP_FIELD(dt_class);
+        CMP_FIELD(mem_type);
+        return true;
+    }
+};
+
+UCS_TEST_P(test_ucp_proto_mock, mock_iface_attr, "NET_DEVICES=mlx5_0:1,lo")
+{
+    setup_mock("ib").set_mock_iface_attr("rc_mlx5/mlx5_0:1",
+        [](uct_iface_attr_t &iface_attr) {
+            iface_attr.dev_num_paths    = 1;
+            iface_attr.cap.am.max_short = 208;
+            iface_attr.bandwidth.shared = 10000000000;
+            iface_attr.latency.c        = 0.000006;
+            iface_attr.latency.m        = 0.000000001;
+        });
+
+    ucp_test::init();
+    if (!has_resource(sender(), "rc_mlx5", "mlx5_0:1")) {
+        UCS_TEST_SKIP_R("no mlx5_0:1 device");
+    }
```

Comment:
can we unite it with mock setup?

Reply:
No, unfortunately we can't..
Because we must setup mock before ucp_init, in order to intercept all the iface opening and perf estimation functions. So we cannot check sender resources at that point.
But I can implement this skip slightly differently, so that we just call a single function here (`check_mocks`) that makes sure ALL the configured mocks got initialized and skip test otherwise, without explicit call to `has_resource`